integration-tests/keycloak-authorization/src/main/resources/application.properties

```diff
@@ -59,6 +59,34 @@ quarkus.keycloak.policy-enforcer.paths.10.enforcement-mode=ENFORCING
 quarkus.keycloak.policy-enforcer.paths.11.path=/api/public-token
 quarkus.keycloak.policy-enforcer.paths.11.enforcement-mode=DISABLED
 
+quarkus.keycloak.policy-enforcer.paths.12.name=Scope Permission Resource
+quarkus.keycloak.policy-enforcer.paths.12.path=/api/permission/annotation/scope-read
+
+quarkus.keycloak.policy-enforcer.paths.13.name=Scope Permission Resource
+quarkus.keycloak.policy-enforcer.paths.13.path=/api/permission/annotation/scope-write
+
+quarkus.keycloak.policy-enforcer.paths.14.name=Scope Permission Resource
+quarkus.keycloak.policy-enforcer.paths.14.path=/api/permission/scopes/programmatic-way
+
+quarkus.keycloak.policy-enforcer.paths.15.name=Scope Permission Resource
+quarkus.keycloak.policy-enforcer.paths.15.path=/api/permission/scopes/standard-way
+quarkus.keycloak.policy-enforcer.paths.15.methods.get.method=GET
+quarkus.keycloak.policy-enforcer.paths.15.methods.get.scopes=read
+
+quarkus.keycloak.policy-enforcer.paths.16.name=Scope Permission Resource
+quarkus.keycloak.policy-enforcer.paths.16.path=/api/permission/scopes/annotation-way
+
+quarkus.keycloak.policy-enforcer.paths.17.name=Scope Permission Resource
+quarkus.keycloak.policy-enforcer.paths.17.path=/api/permission/scopes/standard-way-denied
+quarkus.keycloak.policy-enforcer.paths.17.methods.get.method=GET
+quarkus.keycloak.policy-enforcer.paths.17.methods.get.scopes=write
```

Comment:
@michalvavrik What I'm not sure about is, what the real expectations about this `scopes` property are.

The token may have a scopes property, so I wonder if this is what this property is really about. I.e, if you configured Keycloak to include a scope `print` then if you configure the scope `write` in this test then it should fail. We have to be very clear about it, I start doubting that we can use this property like you propose...

Reply:
np, I acknowledge I have limited knowledge of this topic. But you should ask question (or please kindly explain it to me) why it works exactly as I expected? If I configured scope `write`, than request is only successful when the user has scope `write` and f.e. it fails with `read`. There are tests that prove it. Can you elaborate please?

Reply:
> I.e, if you configured Keycloak to include a scope `print` then if you configure the scope `write` in this test then it should fail. 

I think that's exactly what is happening here, isn't it? Tests I added has in KC `read` and if policy configured in Quarkus requires `write`, it fails. Please see `testPermissionScopes`. Have I misunderstood you?

Reply:
@michalvavrik Can we talk about one specific call there as a lot is going on, so lets select one specific test call and then the questions:
- Tests I added has in KC read - what do you mean here, a `read` scope set in the KC realm, in a specific policy, and is passed to Quarkus in the `scope` token claim ?
- and if policy configured in Quarkus requires write, it fails: but in this case then the selected test should not require the use of `PermissionAllowed` - the token carries the scope value, the configuration says which scope is allowed

@pedroigor Can you clarify please the use of the `scope` Keycloak authorization property ?

@michalvavrik I think your idea is good, but I'm not sure it works for keycloak-authorization. I've been thinking that this is what we have to do at the `quarkus-oidc` level, if a given method has `@PermissionAllowed("read")` then this method can only be accessed if the token has a `read` scope claim

Reply:
Hey @sberyozkin , I appreciate that you always find a time for feedback, I realize the discussion is pretty time consuming. In general, it's important for me to understand why something is rejected, or requested to be changed, so thank you!

>> Tests I added has in KC read
> What do you mean here, a read scope set in the KC realm, in a specific policy, and is passed to Quarkus in the scope token claim ?

`Scope Permission Resource` has scopes `read,write`, however the permission `Scope Permission Resource` only holds scope `read`, therefore it can't access paths with policy that requires `write`.

https://github.com/quarkusio/quarkus/blob/bd513742b1d5d77163dff9ad4b2db067c008a586/integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/KeycloakLifecycleManager.java#L229

In current Quarkus main branch, there is test that ensure this exact behavior. When resource requires only `read, the request succeeds:

https://github.com/quarkusio/quarkus/blob/bd513742b1d5d77163dff9ad4b2db067c008a586/integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/PolicyEnforcerTest.java#L114

When resource requires only `write`, the request fails

https://github.com/quarkusio/quarkus/blob/main/integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/PolicyEnforcerTest.java#L110


> > and if policy configured in Quarkus requires write, it fails
> 
> But in this case then the selected test should not require the use of `PermissionAllowed` - the token carries the scope value, the configuration says which scope is allowed

In the documentation example I added (and tests I added), path that contains `standard-way` does not require `PermissionsAllowed`, and I tried to make it clear as much for users as possible that is what they should use (`standard way`). You can see in configuration properties, that for these paths is set `get.scopes=read`, `get.scopes=write` respectively`:

```
quarkus.keycloak.policy-enforcer.paths.15.name=Scope Permission Resource
quarkus.keycloak.policy-enforcer.paths.15.path=/api/permission/scopes/standard-way
quarkus.keycloak.policy-enforcer.paths.15.methods.get.method=GET
quarkus.keycloak.policy-enforcer.paths.15.methods.get.scopes=read

quarkus.keycloak.policy-enforcer.paths.17.name=Scope Permission Resource
quarkus.keycloak.policy-enforcer.paths.17.path=/api/permission/scopes/standard-way-denied
quarkus.keycloak.policy-enforcer.paths.17.methods.get.method=GET
quarkus.keycloak.policy-enforcer.paths.17.methods.get.scopes=write
```

However there are policies were I didn't set `scopes`, therefore they are not required.

```
quarkus.keycloak.policy-enforcer.paths.14.name=Scope Permission Resource
quarkus.keycloak.policy-enforcer.paths.14.path=/api/permission/scopes/programmatic-way

quarkus.keycloak.policy-enforcer.paths.16.name=Scope Permission Resource
quarkus.keycloak.policy-enforcer.paths.16.path=/api/permission/scopes/annotation-way
```

If your policy does not require `scope`, it doesn't change a thing about the fact that permission `Scope Permission Resource` still holds the scope. I believe the way I use it is correct, because otherwise, why is there existing test that ensures this behavior works?

> @michalvavrik I think your idea is good, but I'm not sure it works for keycloak-authorization. I've been thinking that this is what we have to do at the quarkus-oidc level, if a given method has @PermissionAllowed("read") then this method can only be accessed if the token has a read scope claim

That's no problem at all, if I won't be able to convince you, I'm absolutely happy to drop the idea (or the PR), I've learnt a lot about KC just be writing this examples/docs. But **please**, I'd like to continue discussing this factually till we agree on scope thingy because I'm pretty sure there is a good reason why tests I added are working. At the very least, are you sure users won't _misunderstand_ the docs in a same way as I did? I thing this PR provides improvement on docs side (as now we document something I couldn't reproduce - checking permissions programmatically when the permission is already checked by the Keycloak authorizer)

Reply:
Hi @michalvavrik Np at all, all I'm trying to do here is to figure how this `scope` is verified as I don't yet how. Your description makes sense, but here is a thing, for example, lets say you have only
```
quarkus.keycloak.policy-enforcer.paths.17.path=/api/permission/scopes/standard-way-denied
quarkus.keycloak.policy-enforcer.paths.17.methods.get.method=GET
```

These `path` and `GET` properties can be matched against the request properties, the current HTTP path and verb.

Now as far as `quarkus.keycloak.policy-enforcer.paths.17.methods.get.scopes=read` is concerned, what exactly in the request that it is being matched against ? See what I'm trying to understand ?

I've been thinking that the token should have the `scope` claim, in which case I see how all of your 3 variations work, where in a standard way you set the scope locally, or you use the annotation or the programmatic check is used ?

Can you please print that token you get in the test and paste in `jwt.io` and check the claims ?

If there is no `scope` request property in the token then I don't understand how it works. For example, how does it work in a standard way where you set the `scopes` in the configuration, does `quarkus-keycloak-authorization` posts this value to Keycloak which return the check result, or does it fetch the policy content from Keycloak and compares locally ?

Thanks

Reply:
I verified token on jwt.io, for username `jdoe`, the token contains `realm_access` claim that has roles `user` and `confidential`. That fits realm user created here:
https://github.com/quarkusio/quarkus/blob/main/integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/KeycloakLifecycleManager.java#L152

Now we are getting at the field where I infere behavior from KC/Quarkus code. This is how I read it:

For path `/api/permission` is enforcement enabled, resource permission is granted if resource policy grants access.

https://github.com/quarkusio/quarkus/blob/main/integration-tests/keycloak-authorization/src/main/resources/policies/confidential-policy.js#L2

The resouce `Permission Resource` with path `/api/permission` has policy `confidential-policy` that grants access to identities with realm access `confidential`

https://github.com/quarkusio/quarkus/blob/main/integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/KeycloakLifecycleManager.java#L214

To make it more complicated, it is also possible to specify additional policies for subpaths. So the subpath (AKA resource) `/api/permission/scope` always grants access to any identity through policy

https://github.com/quarkusio/quarkus/blob/main/integration-tests/keycloak-authorization/src/main/resources/policies/always-grant.js#L1

 and scope permission is created that grants to user `read` scope.

https://github.com/quarkusio/quarkus/blob/b97186f9278248d82542656b56f89ac9b5f58765/integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/KeycloakLifecycleManager.java#L289

Now it gets important - policy (confidential policy/always grant policy) grants resource permission (here it is either `Scope Permission Resource` or `Permission Resource`). 

https://github.com/quarkusio/quarkus/blob/b97186f9278248d82542656b56f89ac9b5f58765/integration-tests/keycloak-authorization/src/test/java/io/quarkus/it/keycloak/KeycloakLifecycleManager.java#L271

And resource permission is optionally granted scopes. That means the flow is identity permission (via claim `realm_access`) -> policy -> resource permission. 

The permission checked in the example added is the resource permission.

@sberyozkin

Reply:
@michalvavrik Good stuff, so we see how the token content is linked this configured scope, it is not exactly how I've been thinking about the `scope`, actually, a pretty good explanation is in Curity docs: https://curity.io/resources/learn/scopes-vs-claims/.

So your change in context of the keycloak-authorization is perfect, thanks.
I'd like to encourage you to give it a go and configure Keycloak somehow to include the `scope` claim (after this PR is merged) in the actual token - I suspect it won't make a difference but it is worth investigating what happens if this `scope` claim is present - I think, in the ideal world, if it is present then no any additional configuration would be required (i.e, none of the 3 options you test in this PR) to get the match or mismatch against the policy stored in Keycloak.

But for now it looks good, for `quarkus-oidc` I'll follow up with converting the scope claim to `StringPermission`.
And I guess your follow update where it would be possible to map the Http policy roles to permissions would also work great